pkg/controller/certificates/trigger/policies/gatherer.go

```diff
@@ -31,40 +31,46 @@ import (
 	"github.com/jetstack/cert-manager/pkg/util/predicate"
 )
 
-// Gatherer is used to gather data about a Certificate in order to evaluate
-// its current readiness/state by applying policy functions to it.
-type Gatherer struct {
-	CertificateRequestLister cmlisters.CertificateRequestLister
-	SecretLister             corelisters.SecretLister
-}
-
-func (g *Gatherer) DataForCertificate(ctx context.Context, crt *cmapi.Certificate) (Input, error) {
+// DataForCertificate is used to gather data about a Certificate in order
+// to evaluate its current readiness/state by applying policy functions to
+// it.
+//
+// The returned input.CurrentRevisionRequest and input.Secret are left nil
+// when they cannot be found. The input.Certificate is copied as-is from
+// the given crt.
+func DataForCertificate(ctx context.Context, getSecret func(string) (*v1.Secret, error), lister cmlisters.CertificateRequestNamespaceLister, crt *cmapi.Certificate) (Input, error) {
 	log := logf.FromContext(ctx)
 	// Attempt to fetch the Secret being managed but tolerate NotFound errors.
-	secret, err := g.SecretLister.Secrets(crt.Namespace).Get(crt.Spec.SecretName)
+	secret, err := getSecret(crt.Spec.SecretName)
 	if err != nil && !apierrors.IsNotFound(err) {
 		return Input{}, err
 	}
 
-	// Attempt to fetch the CertificateRequest resource for the current 'status.revision'.
-	var req *cmapi.CertificateRequest
+	// Fetch the CertificateRequest resource for the current
+	// 'status.revision' if it exists; default to using revision "1" since
+	// the issuing controller may be still issuing the first revision of
+	// the certificate.
+	revision := 1
```

Comment:
I don't think we should default this to 1 if it is not set - the `CurrentRevisionRequest` is meant to be the latest *completed* CertificateRequest, not just the latest one that exists. The trigger evaluation functions is then able to make decisions accordingly (i.e. if this is not set, and there is no existing issued Secret AND the Issuing condition is not present, it'll add the condition).

The requestmanager is then responsible for ensuring the current in-progress *certificaterequest* is up to date for the certificate's spec - if it is _not_, it deletes the in-progress CertificateRequest and then creates a new one with the same revision

Reply:
Aaah... I now understand what this "current" CR is all about! This current CR wasn't the one I thought it was 😅

The "current" in `CurrentCertificateRequest` means "the CR associated with the current Certificate", is that correct?